### DISCUSSION TITLE
API's Base structure Idea

### DISCUSSION THREAD

**Discussion by Core-LOVE** [Wohlstand/TheXTech#407] [Ideas]
This is heavily inspired by LunaLua API, but it also does some simplifications and changes. (THIS IS NOT A FULL STRUCTURE OF COURSE!)
1. Lua Modules and related stuff:
- **table**
- **vector**
- **string**
- **math**
- **os**
- **package** (Can also be used to register events and etc)
- **debug**
- **io**
- **net** (Limited socket)

2. Game-related Modules:
- **Assets** (Basically a replacement for Graphics.sprites and other stuff)
- **Graphics**
- **Audio** (Mix of Audio and SFX modules from SMBX2)
- **Level**
- **Episode**
- **Misc**

3. Examples:
Registering events in a simple library:
```lua 
local Library = {}

function Library.onTick()
    print("tick")
end

function Library.onInit() -- basically replaces onInitAPI from SMBX2
   package.registerEvent(Library, 'onTick') -- OR you can just use global registerEvent
end

return Library
```
Registering NPC:
```lua 
local myNPC = {}
local npcID = NPC_ID

NPC.registerDefaultConfig{
	id = npcID,
	
	frames = 0,
}

NPC.registerHarmTypes(
	{
		HARM_TYPE_JUMP = 0.5 -- dmg value
	},
	{
		[HARM_TYPE_JUMP] = 1 -- effect id
	}
) -- unlike SMBX2 you can also define how much damage harm type does. Default value is 1 obviously

function myNPC.onTickNPC(v)
	v.speedY = -1
end

function myNPC.onInit()
	registerNPCEvent(npcID, myNPC, 'onTickNPC') -- OR you can use package.registerNPCEvent
end

return myNPC
```
Drawing image to screen & playing sfx:
```lua
local goombaImage = Assets.npc[1]
local jumpSFX = Assets.sounds[1]

Audio.play(jumpSFX) -- you can also use Audio.play(1) OR playSound(1)

function onDraw()
	Graphics.drawImage(goombaImage, 32, 32)
end
```

**Comment by ds-sloth** [comment]
I'm glad to hear you're interested in helping us construct a scripting API!

I have a few ideas also:
- Since we have control over all of the game's code, we don't need to use `init` and `tick` functions that operate outside of the rest of the game's structure. Instead, we could allow creators to implement each of the unique chunks of NPC logic (activate, hit, update, special, kill, etc) in Lua, or delegate them to the various hardcoded NPCs.
- In order to prevent unexpected difficulties and make scripted code more testable, we could restrict what individual object code is able to do. For instance, let objects use the game's native quadtree capabilities to run collision detection, instead of iterating directly over NPCs, and don't let an object modify another object's state.
- Require NPCs to maintain accurate bounding boxes for rendering, collision, and deactivation purposes. (The render/deactiv one could be larger than the collision one, but this could use the game's existing per-type mechanisms of GFXOffset and GFXWidth.) That way, NPCs will be able to collide with each other accurately, and we will be able to query a quadtree or spatial hash to decide which NPCs to render.
- Carefully consider how to balance flexibility with memory usage. I might be inclined to give scripts global state, but no per-object state beyond the game's six numerical values per NPC and one numerical value per Block. I don't want an overly excited user to allocate ten new strings per block for some block type and then add 1000s of these blocks on their level.

How do those ideas sound to you? I don't have very much working with in-game scripting engines, but I'm trying to balance creativity with performance, as well as the awareness that many of our script creators will not be experienced programmers. I want everyone to be able to have fun with this. I might be interested in making a graphical scripting interface, in the long run, to facilitate this.


One more thing: for performance on a few of our targets and avoiding floating-point implementation differences during NetPlay, we're considering migrating much of the game's numerical logic to fixed-point arithmetic. I'm still not certain whether this will affect compatibility. (C++ doubles already act a tiny bit differently than VB6 Doubles, though the difference is never greater than +/-0.00001 or so.) I also don't know whether we'd want to make this a compile-time option or enable it all of the time, but that would probably depend a lot on performance across platforms and whether there are any compatibility effects.

That means it would be great if we can make the scripting layer as implementation-agnostic as possible while we're still considering this. On most of our platforms it's completely trivial to just cast everything to/from Lua numbers when entering/leaving the scripting API, but I would want to be certain that those Lua numbers act the same on all of our targets. This is also a risk because if we use a 32-bit fixed-point representation internally, rounding errors **will** accumulate during those casts. But somehow, forcing creators to deal with fixed-point math seems a little counter to our accessibility aims, at least from my perspective.

**Comment by 0lhi** [comment]
I have two worries:

1. Performance goals restricting scripting too much, to the point where a lot of cool things can't be done.

2. Scripting not being restricted at all and becoming a bloated monster like SMBX2 or even worse.

Along with the other considerations @ds-sloth mentioned (as well as the possibility of making certain features only available on stronger platforms), it will probably take many in-depth discussions to find a good, balanced approach.

@Core-LOVE If you are interested in contributing to this, would you consider joining the Matrix-Network? Many of the things that are planned for TheXTech have been first discussed by ds, Wohl and myself there. It would be great to have you be there so we can coordinate things better.

**Reply by LoveBodhi** [reply]
How about SMBX-38A's Teascript System??

**Reply by 0lhi** [reply]
Imo not a priority at all, but at #379 @Wohlstand mentioned the possibility of making a Transpiler that would convert TeaScript calls to Lua calls on-the-fly.

**Comment by ds-sloth** [comment]
Thanks to @Core-LOVE for informing me about the SMBX2 architecture on Discord. I can now make a few more concrete suggestions:

(1) Do not use an event registry system. It's more difficult for creators to wrap their heads around, and requires more RAM and processing resources (including in hot loops, for certain events). The only creator-facing advantage of the event registry system is that it allows a single type to have multiple listeners for its events, but this is an intrinsically confusing design (see below).

(2) Instead of having everything be a "library", enforce each script (except level scripts) to have a defined purpose. For instance, `npc-375.lua` will return a structure that defines the logic for NPC 375. All of the event listeners that are specific to NPC 375 should be defined as a part of the structure returned by `npc-375.lua`.
In order to facilitate shared logic, like how Chucks share a lot of their logic in SMBX2, allow a Lua file to include another Lua file in its top line. So, if NPC 375 is a Charging Chuck, it could start with `chargingChuck = import chuck` in Lua syntax.
